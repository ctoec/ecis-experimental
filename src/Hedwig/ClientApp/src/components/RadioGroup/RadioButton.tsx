import React from 'react';

type RadioButtonProps = {
	text: string;
	value: string;
	name: string;
	onClick: (value: any) => any;
	selected?: boolean;
	className?: string;
};

export default function RadioButton({
	text,
	value,
	name,
	onClick,
	selected,
	className,
}: RadioButtonProps) {
	return (
		<div className={`usa-radio ${className}`}>
			<input
				className="usa-radio__input"
				id={value}
				type="radio"
				name={name}
				value={value}
				defaultChecked={selected}
				onClick={value => onClick(value)}
			/>
			<label className="usa-radio__label" htmlFor={value}>
				{text}
			</label>
		</div>
	);
}
