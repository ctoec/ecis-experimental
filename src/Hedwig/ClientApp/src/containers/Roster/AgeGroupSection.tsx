import React from 'react';
import idx from 'idx';
import pluralize from 'pluralize';
import { Link } from 'react-router-dom';
import { Table, TableProps } from '../../components/Table/Table';
import Tag from '../../components/Tag/Tag';
import InlineIcon from '../../components/InlineIcon/InlineIcon';
import { Enrollment, Funding, FundingSpace } from '../../generated';
import { lastFirstNameFormatter } from '../../utils/nameFormatter';
import dateFormatter from '../../utils/dateFormatter';
import getColorForFundingSource, { fundingSourceDetails } from '../../utils/fundingTypeFormatters';
import { DeepNonUndefineable } from '../../utils/types';
import { hasValidationErrors } from '../../utils/validations';
import { tsFilter } from '../../utils/types';

export type AgeGroupTableProps = { id: string; data: DeepNonUndefineable<Enrollment>[] };

type AgeGroupSectionProps = {
	ageGroup: string;
	ageGroupTitle: string;
	enrollments: DeepNonUndefineable<Enrollment>[];
	fundingSpaces?: FundingSpace[];
};

function generateFundingTag(funding: DeepNonUndefineable<Funding>): JSX.Element {
	return (
		<Tag
			key={`${funding.source}-${funding.time}`}
			text={funding.source ? fundingSourceDetails[funding.source].tagFormatter(funding) : ''}
			color={funding.source ? getColorForFundingSource(funding.source) : 'gray-90'}
		/>
	);
}

const defaultRosterTableProps: TableProps<DeepNonUndefineable<Enrollment>> = {
	id: 'roster-table',
	data: [],
	rowKey: row => row.id,
	columns: [
		{
			name: 'Name',
			cell: ({ row }) => (
				<th scope="row">
					<Link to={`/roster/enrollments/${row.id}/`} className="usa-link">
						{lastFirstNameFormatter(row.child)}
					</Link>
					&nbsp;
					{hasValidationErrors(row) ? InlineIcon({ icon: 'incomplete' }) : ''}
				</th>
			),
			sort: row => lastFirstNameFormatter(row.child),
		},
		{
			name: 'Birthdate',
			cell: ({ row }) =>
				(row.child && (
					<td className="oec-table__cell--tabular-nums">
						{row.child.birthdate && dateFormatter(row.child.birthdate)}
					</td>
				)) || <></>,
			sort: row => ((row.child && row.child.birthdate) || new Date(0)).getTime(),
		},
		{
			name: 'Funding',
			cell: ({ row }) => (
				<td>
					{row.fundings && row.fundings.length > 0 ?
						row.fundings.map<React.ReactNode>((funding: DeepNonUndefineable<Funding>) =>
							generateFundingTag(funding)
						) :
						<span className="text-italic text-base">Private pay</span>
					}
				</td>
			),
			sort: row => idx(row, _ => _.fundings[0].source) || '',
		},
		{
			name: 'Enrollment date',
			cell: ({ row }) => (
				<td className="oec-table__cell--tabular-nums">
					{row.entry
						? dateFormatter(row.entry) + (row.exit ? `â€“${dateFormatter(row.exit)}` : '')
						: ''}
				</td>
			),
			sort: row => (row.entry && row.entry.toString()) || '',
		},
	],
	defaultSortColumn: 0,
	defaultSortOrder: 'ascending',
};

export default function AgeGroupSection({
	ageGroup,
	ageGroupTitle,
	enrollments,
	fundingSpaces,
}: AgeGroupSectionProps) {
	if (!enrollments.length) return null;
	const tableProps = Object.assign({}, defaultRosterTableProps, {
		id: `${ageGroup}-roster-table`,
		data: enrollments,
	});

	return (
		<>
			<h2>{`${ageGroupTitle} (${pluralize('child', tableProps.data.length, true)})`}</h2>
			{fundingSpaces && (
				<ul>
					{fundingSpaces.map(space => {
						const enrolledForFunding = tsFilter<Enrollment>(enrollments, e =>
							e.fundings ? tsFilter<Funding>(
								e.fundings, f => f.source === space.source && f.time == space.time
							).length > 0 : false
						).length;

						return (
							<li key={`${space.time}-${ageGroupTitle}`}>
								<span className="text-bold">
									{`${enrolledForFunding}/${space.capacity}`}
								</span>
								<span>
									{` ${(space.time || '').toLowerCase()} time ${(space.source || '')} `}
									spaces filled
								</span>
							</li>
						)
					})}
				</ul>
			)}
			<Table {...tableProps} fullWidth />
		</>
	);
}
