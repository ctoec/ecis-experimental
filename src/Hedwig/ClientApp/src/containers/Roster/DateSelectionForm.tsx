import React, { useState } from 'react';
import moment from 'moment';
import { DatePicker, DateRange } from '../../components/DatePicker/DatePicker';

type DateSelectionFormProps = {
	onSubmit: (newDateRange: DateRange) => any;
	inputDateRange: DateRange;
	byRange: boolean;
};

export default function DateSelectionForm({
	onSubmit,
	inputDateRange,
	byRange,
}: DateSelectionFormProps) {
	const [currentDateRange, setDateRange] = useState<DateRange>(inputDateRange);

	return (
		<DatePicker
			byRange={byRange}
			onChange={newRange => {
				setDateRange(newRange);
				onSubmit(newRange);
			}}
			dateRange={currentDateRange}
			possibleRange={{ startDate: null, endDate: moment().local() }}
		/>
	);
}
