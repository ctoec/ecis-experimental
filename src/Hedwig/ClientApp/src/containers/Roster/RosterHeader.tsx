import React from 'react';
import {
	Button,
	ChoiceList,
	DateInput,
	DateRangeInput,
	DateRange,
	InlineIcon,
} from '../../components';
import getDefaultDateRange from '../../utils/getDefaultDateRange';
import { Site, Organization } from '../../generated';
import { rosterEnrollmentsFormatter } from '../../utils/stringFormatters';
import pluralize from 'pluralize';
import ButtonWithDrowdown from '../../components/ButtonWithDropdown/ButtonWithDrowdown';

type RosterHeaderProps = {
	organization: Organization;
	site?: Site;
	numberOfEnrollments: number;
	showPastEnrollments: boolean;
	toggleShowPastEnrollments: () => void;
	dateRange: DateRange;
	setDateRange: (_: DateRange) => void;
	filterByRange: boolean;
	setFilterByRange: (_: boolean) => void;
};

const RosterHeader: React.FC<RosterHeaderProps> = ({
	organization,
	site,
	numberOfEnrollments,
	showPastEnrollments,
	toggleShowPastEnrollments,
	dateRange,
	setDateRange,
	filterByRange,
	setFilterByRange,
}) => {
	const sites = organization.sites || [];
	const singleSiteProgram = sites.length === 1;

	const handlePastEnrollmentsChange = () => {
		toggleShowPastEnrollments();
		setFilterByRange(false);
		setDateRange(getDefaultDateRange());
	};

	if (!site && !singleSiteProgram) {
		const pluralizedNumKids = pluralize('child', numberOfEnrollments, true);

		return (
			<>
				<div className="grid-row flex-first-baseline flex-space-between">
					<h1 className="tablet:grid-col-auto">{organization.name}</h1>
					<div className="tablet:grid-col-auto">
						<ButtonWithDrowdown
							id="enroll-select"
							className="margin-right-0"
							text="Enroll child"
							dropdownElement={
								<InlineIcon
									icon="arrowDown"
									className="margin-left-1"
									svgProps={{
										fill: 'currentColor',
									}}
								/>
							}
							options={sites.map(s => ({
								text: s.name || '',
								value: `/roster/sites/${s.id}/enroll`,
							}))}
							optionsProps={{
								className: 'position-absolute right-1',
							}}
						/>
					</div>
				</div>
				<div className="margin-bottom-4 grid-row font-sans-lg">
					<div className="tablet:grid-col-fill">
						<div className="intro display-flex flex-row flex-wrap flex-justify-start">
							<span className="display-flex flex-first-baseline">
								{pluralizedNumKids} enrolled across&nbsp;
								<ButtonWithDrowdown
									id="site-select"
									appearance="unstyled"
									className="line-height--small"
									text={<span className="text-bold font-sans-lg"> {sites.length} sites</span>}
									dropdownElement={
										<InlineIcon
											icon="angleArrowDown"
											svgProps={{
												fill: 'currentColor',
											}}
										/>
									}
									options={sites.map(s => ({
										text: s.name || '',
										value: `/roster/sites/${s.id}`,
									}))}
									optionsProps={{
										className: 'font-sans-sm',
									}}
								/>
							</span>
						</div>
					</div>
				</div>
			</>
		);
	} else {
		const numKidsEnrolledText = rosterEnrollmentsFormatter(
			numberOfEnrollments,
			showPastEnrollments,
			dateRange,
			filterByRange
		);

		// If site is undefined, then it is a single site program
		// Assumes there is always at least one site in an organization
		const activeSite = site || sites[0];

		return (
			<>
				<div className="margin-bottom-2 grid-row flex-first-baseline flex-space-between">
					<div className="tablet:grid-col-fill">
						<h1 className="tablet:grid-col-auto">{activeSite.name}</h1>
						<p className="intro display-flex flex-row flex-wrap flex-justify-start">
							{numKidsEnrolledText}
							&nbsp;
							<Button
								text={
									showPastEnrollments ? 'View only current enrollments' : 'View past enrollments'
								}
								appearance="unstyled"
								onClick={handlePastEnrollmentsChange}
							/>
						</p>
					</div>
					<div className="tablet:grid-col-auto">
						<Button
							text="Enroll child"
							href={`/roster/sites/${activeSite.id}/enroll`}
							className="margin-right-0"
						/>
					</div>
				</div>
				{showPastEnrollments && (
					<div className="padding-bottom-2">
						<ChoiceList
							type="radio"
							legend="Select date or date range"
							options={[
								{
									text: 'By date',
									value: 'date',
								},
								{
									text: 'By range',
									value: 'range',
								},
							]}
							onChange={event => setFilterByRange(event.target.value === 'range')}
							horizontal={true}
							id={'dateSelectionType'}
							selected={filterByRange ? ['range'] : ['date']}
							className="margin-top-neg-3"
							// This is goofy but we're getting rid of this soon anyway
						/>
						{filterByRange ? (
							<DateRangeInput
								id="enrollment-roster-datepicker"
								label="Date"
								onChange={newDateRange => (newDateRange ? setDateRange(newDateRange) : null)}
								dateRange={dateRange}
								className="margin-top-neg-3"
							/>
						) : (
							<DateInput
								id="enrollment-roster-datepicker"
								label="Date"
								onChange={newDate => setDateRange({ startDate: newDate, endDate: newDate })}
								date={dateRange.startDate}
								className="margin-top-neg-3"
							/>
						)}
					</div>
				)}
			</>
		);
	}
};

export default RosterHeader;
