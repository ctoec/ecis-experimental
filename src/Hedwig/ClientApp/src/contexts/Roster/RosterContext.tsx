import React, { createContext, useState } from 'react';
import { DateRange } from '../../components';
import getDefaultDateRange from '../../utils/getDefaultDateRange';

export type RosterContextType = {
  showPastEnrollments: boolean;
  toggleShowPastEnrollments: () => void;
  dateRange: DateRange;
  setDateRange: (_: DateRange) => void;
  filterByRange: boolean;
  setFilterByRange: (_: boolean) => void;

};

const RosterContext = createContext<RosterContextType>({
  showPastEnrollments: false,
  toggleShowPastEnrollments: () => {},
  dateRange: getDefaultDateRange(),
  setDateRange: (_) => {},
  filterByRange: false,
  setFilterByRange: (_) => {},
});

const { Provider, Consumer } = RosterContext;

export type RosterProviderPropsType = {};

const RosterProvider: React.FC<RosterProviderPropsType> = ({ children }) => {
	const [showPastEnrollments, toggleShowPastEnrollments] = useState(false);
	const [dateRange, setDateRange] = useState<DateRange>(getDefaultDateRange());
  const [filterByRange, setFilterByRange] = useState(false);
  
	return (
		<Provider value={{
      showPastEnrollments,
      toggleShowPastEnrollments: () => toggleShowPastEnrollments(!showPastEnrollments),
      dateRange,
      setDateRange,
      filterByRange,
      setFilterByRange
    }}>
			{children}
		</Provider>
	);
};

export { RosterProvider };
export { Consumer as RosterConsumer };
export default RosterContext;