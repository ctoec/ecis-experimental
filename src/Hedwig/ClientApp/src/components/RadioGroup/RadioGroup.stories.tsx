import React from 'react';
import { storiesOf } from '@storybook/react';
import { action } from '@storybook/addon-actions';
import RadioGroup from './RadioGroup';

const onClick = action('onClick');
const options = [
	{
		selected: true,
		text: 'Option one',
		value: 'one',
	},
	{
		selected: false,
		text: 'Option two',
		value: 'two',
	},
];

storiesOf('RadioGroup', module)
	.add('Default', () => {
		return (
			<RadioGroup
				options={options}
				groupName="storybook-radio-group"
				onClick={onClick}
				legend="Storybook radio buttons"
			/>
		);
	})
	.add('Horizontal', () => {
		return (
			<RadioGroup
				options={options}
				groupName="storybook-radio-group-2"
				onClick={onClick}
				legend="Storybook radio buttons"
				horizontal={true}
			/>
		);
	})
	.add('Default selection', () => {
		return (
			<RadioGroup
				options={options}
				groupName="storybook-radio-group-default-selection"
				onClick={onClick}
				legend="Storybook radio buttons"
				selected="one"
			/>
		);
	})
	.add('With error', () => {
		return (
			<RadioGroup
				options={options}
				groupName="storybook-radio-group-with-error"
				onClick={onClick}
				legend="Storybook radio buttons"
				showError={true}
				errorMessage="You  must select one or two."
			/>
		);
	});
