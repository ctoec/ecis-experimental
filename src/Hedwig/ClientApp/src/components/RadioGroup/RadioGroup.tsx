import React from 'react';
import RadioButton from './RadioButton';
import FieldSet from '../FieldSet/FieldSet';
import Legend from '../Legend/Legend';

type RadioButtonOptions = {
	text: string;
	value: string;
};

type RadioGroupProps = {
	options: RadioButtonOptions[];
	groupName: string;
	onChange: (event: React.ChangeEvent<HTMLInputElement>) => any;
	selected?: string;
	horizontal?: boolean;
	legend: string;
	warning?: string;
};

export default function RadioGroup({
	options,
	groupName,
	onChange,
	selected,
	horizontal,
	legend,
	warning
}: RadioGroupProps) {
	return (
		<FieldSet
			legend={legend}
			warning={warning}
		>
			<div className={horizontal ? 'grid-row flex-align-start grid-gap' : ''}>
				{options.map(option => (
					<RadioButton
						value={option.value}
						text={option.text}
						onChange={onChange}
						key={option.value}
						name={groupName}
						selected={option.value === selected}
						className={horizontal ? 'grid-col flex-auto' : ''}
					/>
				))}
			</div>
		</FieldSet>

	);
}
