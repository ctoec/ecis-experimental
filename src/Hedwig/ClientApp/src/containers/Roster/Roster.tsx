import React, { useState, useContext } from 'react';
import { Link } from 'react-router-dom';
import idx from 'idx';
import nameFormatter from '../../utils/nameFormatter';
import dateFormatter from '../../utils/dateFormatter';
import enrollmentTextFormatter from '../../utils/enrollmentTextFormatter';
import getDefaultDateRange from '../../utils/getDefaultDateRange';
import getColorForFundingSource, { fundingSourceDetails } from '../../utils/getColorForFundingType';
import { Table, TableProps } from '../../components/Table/Table';
import Tag from '../../components/Tag/Tag';
import { DateRange } from '../../components/DatePicker/DatePicker';
import Button from '../../components/Button/Button';
import RadioGroup from '../../components/RadioGroup/RadioGroup';
import Legend from '../../components/Legend/Legend';
import useApi from '../../hooks/useApi';
import UserContext from '../../contexts/User/UserContext';
import { Enrollment } from '../../OAS-generated/models/Enrollment';
import DateSelectionForm from './DateSelectionForm';

export default function Roster() {
	const [showPastEnrollments, toggleShowPastEnrollments] = useState(false);
	const [dateRange, setDateRange] = useState<DateRange>(getDefaultDateRange());
	const [byRange, setByRange] = useState(false);
	const { user } = useContext(UserContext);
	const [loading, error, site] = useApi(
		api =>
			api.apiOrganizationsOrgIdSitesIdGet({
				// TODO after pilot: don't just grab the first org and site
				orgId: idx(user, _ => _.orgPermissions[0].organization.id) || 0,
				id: idx(user, _ => _.orgPermissions[0].organization.sites[0].id) || 0,
			}),
		[user]
  );
  
  	const [enrollmentsLoading, enrollmentsError, rawEnrollments] = useApi(
			// TODO: after everything being nullable is solved, ditch raw enrollments and type mapping below
			api =>
				api.apiOrganizationsOrgIdSitesSiteIdEnrollmentsGet({
					// TODO after pilot: don't just grab the first org and site
					orgId: idx(user, _ => _.orgPermissions[0].organization.id) || 0,
					siteId: idx(user, _ => _.orgPermissions[0].organization.sites[0].id) || 0,
					include: ['child', 'fundings'],
					startDate:
						(dateRange && dateRange.startDate && dateRange.startDate.toDate()) ||
						undefined,
					endDate:
						(dateRange && dateRange.endDate && dateRange.endDate.toDate()) ||
						undefined,
				}),
			[user, dateRange]
		);

	function handlePastEnrollmentsChange() {
		toggleShowPastEnrollments(!showPastEnrollments);
		setByRange(false);
		setDateRange(getDefaultDateRange());
	}

	if (!site) {
		return <div className="Roster"></div>;
  }
  
	// TODO: FIX THIS-- ditch raw enrollments
	const enrollments = (rawEnrollments || []).map(e => e as Required<Enrollment>);

	const rosterTableProps: TableProps<Required<Enrollment>> = {
		id: 'roster-table',
		data: enrollments,
		rowKey: row => row.id,
		columns: [
			{
				name: 'Name',
				cell: ({ row }) => (
					<th scope="row">
						<Link to={`/roster/enrollments/${row.child.id}/`} className="usa-link">
							{nameFormatter(row.child)}
						</Link>
					</th>
				),
				sort: row => nameFormatter(row.child),
			},
			{
				name: 'Date of birth',
				cell: ({ row }) => (
					<td className="oec-table__cell--tabular-nums">
						{row.child.birthdate && dateFormatter(row.child.birthdate)}
					</td>
				),
				sort: row => (row.child.birthdate || new Date(0)).getTime(),
			},
			{
				name: 'Funding',
				cell: ({ row }) => (
					<td>
						{row.fundings && row.fundings.length ? (
							<Tag
								text={`${row.fundings[0].source}`}
								color={
									row.fundings[0].source
										? getColorForFundingSource(row.fundings[0].source)
										: 'gray-90'
								}
							/>
						) : (
							''
						)}
					</td>
				),
			},
			{
				name: 'Enrolled',
				cell: ({ row }) => (
					<td className="oec-table__cell--tabular-nums">
						{row.entry
							? dateFormatter(row.entry) + 'â€“' + (row.exit ? dateFormatter(row.exit) : '')
							: ''}
					</td>
				),
				sort: row => (row.entry && row.entry.toString()) || '',
			},
		],
		defaultSortColumn: 0,
		defaultSortOrder: 'asc',
	};

	const numKidsEnrolledText = enrollmentTextFormatter(
		enrollments.length,
		showPastEnrollments,
		dateRange,
		byRange
	);

	const legendItems = Object.keys(fundingSourceDetails).map(key => ({
		text: fundingSourceDetails[key].fullTitle,
		symbolColor: fundingSourceDetails[key].colorToken,
		number: enrollments.filter(
			(row: Required<Enrollment>) =>
				row.fundings && row.fundings.filter((funding: any) => funding.source === key).length > 0
		).length,
	}));

	return (
		<div className="Roster">
			<section className="grid-container">
				<h1 className="grid-col-auto">{site.name}</h1>
				<Legend items={legendItems} />
				<div className="grid-row">
					<div className="tablet:grid-col-fill">
						<p className="usa-intro display-flex flex-row flex-wrap flex-justify-start">
							<span className="margin-right-2 flex-auto">{numKidsEnrolledText}</span>
							<Button
								text={
									showPastEnrollments ? 'Show only current enrollments' : 'Show past enrollments'
								}
								appearance="unstyled"
								onClick={handlePastEnrollmentsChange}
							/>
						</p>
					</div>
					<div className="tablet:grid-col-auto">
						<Button text="Enroll child" href={`/roster/sites/${site.id}/enroll`} />
					</div>
				</div>
				{showPastEnrollments && (
					<div className="usa-fieldset">
						<RadioGroup
							options={[
								{
									text: 'By date',
									value: 'date',
								},
								{
									text: 'By range',
									value: 'range',
								},
							]}
							onChange={event => setByRange(event.target.value === 'range')}
							horizontal={true}
							groupName={'dateSelectionType'}
							legend="Select date or date range."
							selected={byRange ? 'range' : 'date'}
						/>
						<DateSelectionForm
							inputDateRange={dateRange}
							byRange={byRange}
							onReset={() => {
								setByRange(false);
								setDateRange(getDefaultDateRange());
							}}
							onSubmit={(newDateRange: DateRange) => setDateRange(newDateRange)}
						/>
					</div>
				)}
				<Table {...rosterTableProps} fullWidth />
			</section>
		</div>
	);
}
