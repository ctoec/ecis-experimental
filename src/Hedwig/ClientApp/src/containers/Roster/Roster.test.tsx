import React from 'react';
import moment from 'moment';
import { mount } from 'enzyme';
import { BrowserRouter } from 'react-router-dom';
import { MockedProvider } from '@apollo/react-testing';
import 'react-dates/initialize';
import Roster, { ROSTER_QUERY } from './Roster';
import DateSelectionForm from './DateSelectionForm';
import RadioGroup from '../../components/RadioGroup/RadioGroup';

const earlierFakeEnrollment = {
	child: {
		firstName: 'James',
		middleName: 'Sirius',
		lastName: 'Potter',
		birthdate: '2015-04-28',
		suffix: null,
	},
	entry: '2019-01-01',
	exit: null,
	fundings: [],
	id: 1,
};
const laterFakeEnrollment = {
	child: {
		firstName: 'Lily',
		middleName: 'Luna',
		lastName: 'Potter',
		birthdate: '2016-12-12',
		suffix: null,
	},
	entry: '2019-03-03',
	exit: null,
	fundings: [
		{
			entry: '2019-03-01',
			exit: '2019-04-01',
			source: 'Gringotts Goblins Grant for Gifted Girls',
		},
	],
	id: 2,
};

const mocks = [
	{
		request: {
			query: ROSTER_QUERY,
			variables: {
				from: moment()
					.local()
					.format('YYYY-MM-DD'),
				to: moment()
					.local()
					.format('YYYY-MM-DD'),
			},
		},
		result: {
			data: {
				me: {
					sites: [
						{
							id: 1,
							name: 'Site 1',
							enrollments: [earlierFakeEnrollment, laterFakeEnrollment],
						},
					],
				},
			},
		},
	},
	{
		request: {
			query: ROSTER_QUERY,
			variables: { from: '2018-01-01', to: '2019-02-01' },
		},
		result: {
			data: {
				me: {
					sites: [
						{
							id: 1,
							name: 'Site 1',
							enrollments: [earlierFakeEnrollment],
						},
					],
				},
			},
		},
	},
];

const waitForUpdate = async (wrapper: any) => {
	await new Promise(resolve => setTimeout(resolve, 10));
	wrapper.update();
};

describe('Roster', () => {
	it('matches snapshot', () => {
		const wrapper = mount(
			<MockedProvider mocks={mocks}>
				<BrowserRouter>
					<Roster />
				</BrowserRouter>
			</MockedProvider>
		);
		expect(wrapper).toMatchSnapshot();
		wrapper.unmount();
	});

	it('renders intro text with the correct number of kids', async () => {
		const wrapper = mount(
			<MockedProvider mocks={mocks} addTypename={false}>
				<BrowserRouter>
					<Roster />
				</BrowserRouter>
			</MockedProvider>
		);
		await waitForUpdate(wrapper);
		const introSpanText = wrapper.find('.usa-intro span').text();
		expect(introSpanText).toBe('2 kids enrolled.');
		wrapper.unmount();
	});

	it('updates the number of kids', async () => {
		const wrapper = mount(
			<MockedProvider mocks={mocks} addTypename={false}>
				<BrowserRouter>
					<Roster />
				</BrowserRouter>
			</MockedProvider>
		);
		await waitForUpdate(wrapper);
		wrapper
			.find('Button')
			.props()
			.onClick();
		await waitForUpdate(wrapper);

		const radioGroup = wrapper.find(RadioGroup);
		radioGroup.props().onClick('range');
		await waitForUpdate(wrapper);

		wrapper
			.find(DateSelectionForm)
			.props()
			.onSubmit({ startDate: moment('2018-01-01'), endDate: moment('2019-02-01') });
		await waitForUpdate(wrapper);

		const introSpanText = wrapper.find('.usa-intro span').text();
		expect(introSpanText).toBe('1 kid was enrolled between January 1, 2018 and February 1, 2019.');
		wrapper.unmount();
	});
});
