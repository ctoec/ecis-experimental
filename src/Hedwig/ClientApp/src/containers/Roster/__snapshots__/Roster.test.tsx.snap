// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Roster matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="grid-container"
  />
  <div
    class="grid-container"
  >
    <div
      class="grid-row flex-first-baseline flex-space-between"
    >
      <h1
        class="tablet:grid-col-auto"
      >
        Children's Adventure Center
      </h1>
      <div
        class="tablet:grid-col-auto"
      >
        <a
          class="usa-button margin-right-0"
          href="/roster/sites/1/enroll"
        >
          Enroll child
        </a>
      </div>
    </div>
    <div
      class="grid-row"
    >
      <div
        class="tablet:grid-col-fill"
      >
        <p
          class="intro display-flex flex-row flex-wrap flex-justify-start"
        >
          <span
            class="margin-right-2 flex-auto"
          >
            5 children enrolled.
          </span>
          <button
            class="usa-button usa-button--unstyled"
            type="button"
          >
            Filter for past enrollments
          </button>
        </p>
      </div>
    </div>
    <div
      class="grid-row flex-wrap margin-y-2 grid-gap oec-legend"
    >
      <div
        class="margin-right-1"
      >
        <div
          class="oec-legend__symbol undefined"
        >
          <span
            class="usa-tag position-relative top-neg-2px bg-blue-50v"
          >
            CDC
          </span>
        </div>
        <div
          class="width-fit-content display-inline margin-left-1 "
        >
          <span
            class="text-bold"
          >
            5/0
          </span>
          <span>
             Child Day Care spaces filled
          </span>
        </div>
      </div>
      <div
        class="margin-right-1"
      >
        <div
          class="oec-legend__symbol undefined"
        >
          <span
            class="oec-inline-icon oec-inline-icon--incomplete"
          >
            <svg>
              error.svg
            </svg>
            <span
              class="usa-sr-only"
            >
              (incomplete)
            </span>
          </span>
        </div>
        <div
          class="width-fit-content display-inline margin-left-1 "
        >
          <span
            class="text-bold"
          >
            3
          </span>
          <span>
             missing information
          </span>
        </div>
      </div>
    </div>
    <h2
      class="margin-top-6"
    >
      Preschool (5 children)
    </h2>
    <ul />
    <table
      class="oec-table oec-table--full-width"
      id="Preschool-roster-table"
    >
      <thead>
        <tr>
          <th
            aria-sort="ascending"
            class="oec-table__column-header oec-sortable"
            role="columnheader"
            scope="col"
            style="width: 35%;"
          >
            <button
              aria-label="Sort table by Name in descending order"
              class="oec-table__column-title usa-button--unstyled width-full"
            >
              Name
              <div
                class="oec-table__sort-controls"
              >
                <span
                  class="oec-table__sort-controls__asc active"
                />
                <span
                  class="oec-table__sort-controls__desc"
                />
              </div>
            </button>
          </th>
          <th
            aria-sort="ascending"
            class="oec-table__column-header oec-sortable"
            role="columnheader"
            scope="col"
            style="width: 20%;"
          >
            <button
              aria-label="Sort table by Birthdate in ascending order"
              class="oec-table__column-title usa-button--unstyled width-full"
            >
              Birthdate
              <div
                class="oec-table__sort-controls"
              >
                <span
                  class="oec-table__sort-controls__asc"
                />
                <span
                  class="oec-table__sort-controls__desc"
                />
              </div>
            </button>
          </th>
          <th
            aria-sort="ascending"
            class="oec-table__column-header oec-sortable"
            role="columnheader"
            scope="col"
            style="width: 25%;"
          >
            <button
              aria-label="Sort table by Funding in ascending order"
              class="oec-table__column-title usa-button--unstyled width-full"
            >
              Funding
              <div
                class="oec-table__sort-controls"
              >
                <span
                  class="oec-table__sort-controls__asc"
                />
                <span
                  class="oec-table__sort-controls__desc"
                />
              </div>
            </button>
          </th>
          <th
            aria-sort="ascending"
            class="oec-table__column-header oec-sortable"
            role="columnheader"
            scope="col"
            style="width: 20%;"
          >
            <button
              aria-label="Sort table by Enrollment date in ascending order"
              class="oec-table__column-title usa-button--unstyled width-full"
            >
              Enrollment date
              <div
                class="oec-table__sort-controls"
              >
                <span
                  class="oec-table__sort-controls__asc"
                />
                <span
                  class="oec-table__sort-controls__desc"
                />
              </div>
            </button>
          </th>
        </tr>
      </thead>
      <tbody>
        <tr>
          <th
            scope="row"
          >
            <a
              class="usa-link print"
              href="/roster/sites/1/enrollments/1/"
            >
              Potter, Lily Luna
            </a>
             
          </th>
          <td
            class="oec-table__cell--tabular-nums"
          >
            12/12/2016
          </td>
          <td>
            <span
              class="usa-tag undefined bg-blue-50v"
            >
              CDC–FT
            </span>
          </td>
          <td
            class="oec-table__cell--tabular-nums"
          >
            02/03/2018
          </td>
        </tr>
        <tr>
          <th
            scope="row"
          >
            <a
              class="usa-link print"
              href="/roster/sites/1/enrollments/2/"
            >
              Potter, Lily Luna
            </a>
             
            <span
              class="oec-inline-icon oec-inline-icon--incomplete"
            >
              <svg>
                error.svg
              </svg>
              <span
                class="usa-sr-only"
              >
                (incomplete)
              </span>
            </span>
          </th>
          <td
            class="oec-table__cell--tabular-nums"
          >
            12/12/2016
          </td>
          <td>
            <span
              class="usa-tag undefined bg-blue-50v"
            >
              CDC–FT
            </span>
          </td>
          <td
            class="oec-table__cell--tabular-nums"
          >
            02/03/2018
          </td>
        </tr>
        <tr>
          <th
            scope="row"
          >
            <a
              class="usa-link print"
              href="/roster/sites/1/enrollments/4/"
            >
              Potter, Lily Luna
            </a>
             
            <span
              class="oec-inline-icon oec-inline-icon--incomplete"
            >
              <svg>
                error.svg
              </svg>
              <span
                class="usa-sr-only"
              >
                (incomplete)
              </span>
            </span>
          </th>
          <td
            class="oec-table__cell--tabular-nums"
          >
            12/12/2016
          </td>
          <td>
            <span
              class="usa-tag undefined bg-blue-50v"
            >
              CDC–FT
            </span>
          </td>
          <td
            class="oec-table__cell--tabular-nums"
          >
            02/03/2018
          </td>
        </tr>
        <tr>
          <th
            scope="row"
          >
            <a
              class="usa-link print"
              href="/roster/sites/1/enrollments/5/"
            >
              Potter, Lily Luna
            </a>
             
          </th>
          <td
            class="oec-table__cell--tabular-nums"
          >
            12/12/2016
          </td>
          <td>
            <span
              class="usa-tag undefined bg-blue-50v"
            >
              CDC–FT
            </span>
          </td>
          <td
            class="oec-table__cell--tabular-nums"
          >
            03/01/2019
          </td>
        </tr>
        <tr>
          <th
            scope="row"
          >
            <a
              class="usa-link print"
              href="/roster/sites/1/enrollments/3/"
            >
              Potter, undefined Luna
            </a>
             
            <span
              class="oec-inline-icon oec-inline-icon--incomplete"
            >
              <svg>
                error.svg
              </svg>
              <span
                class="usa-sr-only"
              >
                (incomplete)
              </span>
            </span>
          </th>
          <td
            class="oec-table__cell--tabular-nums"
          >
            12/12/2016
          </td>
          <td>
            <span
              class="usa-tag undefined bg-blue-50v"
            >
              CDC–FT
            </span>
          </td>
          <td
            class="oec-table__cell--tabular-nums"
          >
            02/03/2018
          </td>
        </tr>
      </tbody>
    </table>
  </div>
</DocumentFragment>
`;
